### PR TITLE
Jesse Baxter

### PR DESCRIPTION
**Pull request by jbaxterUO**

### CODE REVIEW

File: brevets/templates/calc.html [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+                              //Fill in each row so long as there is information for each row
+                              $('.control').each(function(index){
+                                 if(index < num_checkpoints){
+                                    var row = $(this);
+                                    row.find("input[name='location']").val(checkpoints[index].location);
+                                    row.find("input[name='open']").val(checkpoints[index].start_time);
+                                    row.find("input[name='close']").val(checkpoints[index].close_time);
+                                    row.find("input[name='km']").val(checkpoints[index].distance);
+                                    
+                                    //This just fills the miles back in incase they are wanted
```
Comment (vlanier2) [inline brevets/templates/calc.html:278]: when you populate the rows back into the webpage you can trigger the change something like this:

$(this).find('input[name="km"]').trigger('change');

This will cause the miles to recalculate as if it was entered by hand.

File: brevets/templates/calc.html [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+            <!-- row -->
+         </form>
+         <script type="text/javascript">
+            var SCRIPT_ROOT = {{ request.script_root|tojson|safe }} ;
+            var TIME_CALC_URL = SCRIPT_ROOT + "/_calc_times";
+            var SUBMIT_URL = SCRIPT_ROOT + "/_submit";
+            var DISPLAY_URL = SCRIPT_ROOT + "/_display";
+
+
+            function clear_inputs(){
```
Comment (vlanier2) [inline brevets/templates/calc.html:126]: Really nice helper function to clear input

File: brevets/templates/calc.html [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+                     //Clear input fields after values have been gathered
+                     clear_inputs();
+
+                     //Add everything together into our brevet
+                     var brevet = {
+                        "length": length,
+                        "start_time" : start_time,
+                        "checkpoints" : checkpoints,
+                     };
+
```
Comment (vlanier2) [inline brevets/templates/calc.html:227]: I found this very readable and well structured. You added great comments, but I think with the spacing and clarity of your layout you hardly needed the comments. Particularly the clean buildup of each checkpoint and the final brevet construction is very easy to read.

File: brevets/tests/test_acp_times.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
+"""
+Nose tests for acp_times.py
```
Comment (vlanier2) [inline brevets/tests/test_acp_times.py:2]: I liked the clear names of each testcase. I can tell immediately the use/edge case that each test is handling. I see a test for the low distance checkpoints and for checkpoints over the total brevet length.

File: brevets/acp_times.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+set_end_tims = {
+         200: 13.5,
+         300: 20, 
+         400: 27,
+         600: 40,
+         1000: 75
+      }
+
+#Known brevet distances
+brev_distances = [0, 200, 400, 600, 1000]
```
Comment (vlanier2) [inline brevets/acp_times.py:46]: Each review I've seen does the tables a bit differently, but I think splitting them up like this makes the most readable code later on.

**Review by vlanier2 — COMMENTED** [review]
Hi Jesse,

Overall the project worked great. The webpage worked as expected and the API worked when tested a bit by hand. I thought your code was particularly concise and logical. Particularly some of the work with JS was very clear and readable. 

The success/error messages at the top of the page was a great addition. I would suggest having that message clear after making further edits to the page, or when hitting the clear button. Also, it reads "Successful insertion into database." when I tried to submit an empty brevet, but I could tell the brevet was not actually inserted and it didnt overwrite the last real brevet I entered. It seems like a couple more unique error messages would be good.

The README was brief, but it told me what I needed to start the project and test without needing to look at your compose file. 

I noticed you used string fields to avoid converting to and from ISO date format (I think). I'm not sure if this was ok, but the schema suggested datetime fields which require python datetime objects. I do think this implementation makes some of the code shorter and maybe makes more sense for the project. 

Some return messages on the API for PUT and DELETE would be nice for testing by hand. Maybe just a success confirmation and the ID. 

The whole code was focused and as concise as I've seen, even the ACP times calculation. Comments were clear and not overused in my opinion. 

Best of luck next term.
